Prometheus and Grafana Integration

alexstrive opened the pull request:
https://www.notion.so/5827b750cd944535883639e5b2139ad1

alexstrive committed:
Batch commit

alexstrive committed:
declare exception metrics

alexstrive committed:
change name retrieval method

alexstrive committed:
integrate metric with custom exceptions

alexstrive committed:
declare kafka specific metrics

alexstrive committed:
integrate previously declared metrics for kafka

andrsuh commented on prometheus.yml:
```diff
@@ -0,0 +1,16 @@
+      - targets: ['localhost:9090']
+
+  - job_name: 'choreographer'
+    metrics_path: '/actuator/prometheus'
```
этот пусть актуален при отправке не через micrometer?

alexstrive replied:
Я предполагаю, что ты хотел написать "путь".

Нет, данный путь предназначен только для метрик, сгенерированными Micrometer.
Путь может быть только один у объявленного job. Так как там больше информации собирается: данные о JVM, CPU и так далее, то выбрал его.

andrsuh replied:
ну я к этому и клюню, нужен коммент и дока

andrsuh commented on prometheus.yml:
```diff
@@ -0,0 +1,16 @@
+    scrape_interval: 5s
+
+    static_configs:
+      - targets: ['localhost:9090']
```
нужно оооочень подробные комменты ко всем пропертям. И здесь и в виде документации в notion.

andrsuh commented on prometheus.yml:
```diff
@@ -0,0 +1,16 @@
+global:
```
этот файл конфигурирует сам прометей, нужно написать доку по поводу того, что этот файл не помещается в проект, а должен быть прокинут в контейнер с прометеем. Нужно создать отдельную репу для хранения и изменения этого файла, поскольку сервисов будет больше. Туда же можно поместить композ для прометея и графаны. Репу создавай в этом же спейсе

andrsuh replied:
все должно быть отмечено в документации на отдельной страничку в notion

alexstrive replied:
"Репу создавай в этом же спейсе" – в каком конкретно? В https://github.com/microservices-course-itmo или под своим именем? Если первое, то у меня нет прав, так как я не участник.

Я так понимаю, что я оставляю исключительно Prometheus и Grafana в docker-compose в этом репозитории?

andrsuh replied:
второй вопрос не понял

andrsuh replied:
права сейчас выдам

alexstrive replied:
Сейчас в docker-compose 4 сервиса: Zookeeper, Kafka, Prometheus, Grafana. Мне стоит последние два (Prometheus, Grafana) вынести в отдельный docker-compose? Таким образом остается 2 в этом репозитории и 2 в новом.

andrsuh replied:
https://github.com/microservices-course-itmo/metrics

andrsuh replied:
да, все как ты написал

alexstrive replied:
`remote: Permission to microservices-course-itmo/metrics.git denied to alexstrive.`

Не могу отправить комит в репозиторий. Могу через pull request оформить изменения.

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/SagaMetrics.java:
```diff
@@ -0,0 +1,133 @@
+import io.prometheus.client.Summary;
+
+/**
+ * This Class expose methods for recording specific metrics
```
Очень хорошая и лаконичная дока! Спасибо

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/SagaMetrics.java:
```diff
@@ -0,0 +1,133 @@
+        prometheusSagaTemplateStepExecutedSummary.labels(sagaName, stepName).observe(timeExecution);
+    }
+
+    public static void incrementSagaTemplate(String sagaName) {
```
подумай, может быть, лучше `incrementSagaTemplate`

alexstrive replied:
Метод ведь так и называется?

andrsuh replied:
лучше count хочет написать, извиняюсь

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/SagaMetrics.java:
```diff
@@ -0,0 +1,133 @@
+ *
+ * TODO: consider better namings for class name and name methods
+ */
+public abstract class SagaMetrics {
```
нет никакой предпосылки делать класс абстрактным

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/SagaMetrics.java:
```diff
@@ -0,0 +1,133 @@
+            .register();
+
+
+    public static void recordSagaInstanceStep(String sagaName, String stepName, double timeExecution) {
```
методы не делаем статическими. И используем DI для внедрения инстанса в нужные классы, пожалуйста

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/components/PrometheusMetricsInjector.java:
```diff
@@ -0,0 +1,17 @@
+public class PrometheusMetricsInjector {
+    @Bean
+    public ServletRegistrationBean injectPrometheusServlet() {
+        var targetUrl = "/metrics-prometheus";
```
плиз доку подробнее

andrsuh replied:
например, что за урл, что за метод, для чего

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/components/SagaHelper.java:
```diff
@@ -0,0 +1,55 @@
+        var sagaTemplateNames = getSagaTemplateNames();
+
+        for (var sagaTemplateName : sagaTemplateNames) {
+            SagaMetrics.incrementSagaTemplate(sagaTemplateName);
```
DI

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/exception/FormattedRuntimeException.java:
```diff
@@ -1,13 +1,16 @@
 public class FormattedRuntimeException extends RuntimeException {
     public FormattedRuntimeException(String messagePattern, Object... params) {
         super(MessageFormatter.arrayFormat(messagePattern, params).getMessage());
+        SagaMetrics.incrementCoordinatorExceptionsThrown(this);
```
Ну нет, не все исключения стоит слать. Тем более, что они часто заворачиваются друг в друга и мы получим кучу инкрементов, когда хотим увидеть один главный

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/kafka/KafkaClient.java:
```diff
@@ -117,6 +118,9 @@ private void startListeningTopics() {
                         } catch (Exception e) {
                             log.error("Error while handling event", e);
                         }
+                        finally {
```
форматирование finally не надо с новой строки

andrsuh commented on src/main/resources/application.properties:
```diff
@@ -3,3 +3,9 @@ spring.data.neo4j.password=secret
 spring.kafka.consumer.group-id=supervisor
+
+
+management.endpoints.web.exposure.include=*
```
может, не стоит все открывать?

andrsuh reviewed (commented):
Все круто, но есть несколько коментов)

alexstrive committed:
close not-used actuator's paths

alexstrive committed:
formatting amends

alexstrive committed:
add documentation for metrics injector

alexstrive committed:
https://github.com/microservices-course-itmo/saga/pull/6#discussion_r444180544
https://github.com/microservices-course-itmo/saga/pull/6#discussion_r444180211
https://github.com/microservices-course-itmo/saga/pull/6#discussion_r444181385

alexstrive committed:
remove metric recording from FormattedRuntimeException

alexstrive committed:
extract Prometheus and Grafana from docker-compose
revert to initial state

alexstrive committed:
remove prometheus configuration file

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/components/SagaMetrics.java:
```diff
@@ -0,0 +1,135 @@
+ * This Class expose methods for recording specific metrics
+ * It changes metrics of Micrometer and Prometheus simultaneously
+ * Micrometer's metrics exposed at /actuator/prometheus
+ * Prometheus' metrics exposed at /metrics-prometheus
```
`{@link PrometheusMetricsInjector#injectPrometheusServlet}`

alexstrive replied:
Добавлено. Метод дополнен документацией.

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/components/SagaMetrics.java:
```diff
@@ -0,0 +1,135 @@
+        prometheusSagaInstanceCompensatedCounter.labels(sagaName).inc();
+    }
+
+    public void incrementCoordinatorExceptionsThrown(Exception exception) {
```
может быть, count?

alexstrive replied:
Изменено.

andrsuh replied:
не вижу почему-то. Предлагаю сделать все методы с префиксом count

andrsuh replied:
?

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/exception/FormattedException.java:
```diff
@@ -1,17 +1,23 @@
 
     public FormattedException(String messagePattern, Object... params) {
         super(MessageFormatter.arrayFormat(messagePattern, params).getMessage());
+        sagaMetrics.incrementCoordinatorExceptionsThrown(this);
```
я уже объяснял почему считаю это неправильным

alexstrive replied:
Удалено

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/exception/KafkaRuntimeException.java:
```diff
@@ -1,11 +1,18 @@
 
     public KafkaRuntimeException(String messagePattern, Throwable cause, Object... params) {
         super(messagePattern, cause, params);
+        sagaMetrics.incrementCoordinatorExceptionsThrown(this);
```
про exceptions уже тоже писал

alexstrive replied:
Удалено

andrsuh commented on src/main/java/com/microservices/saga/choreography/supervisor/exception/StepDefinitionNotFoundException.java:
```diff
@@ -11,5 +16,6 @@
      */
     public StepDefinitionNotFoundException(String messagePattern, Object... params) {
         super(messagePattern, params);
+        sagaMetrics.incrementCoordinatorExceptionsThrown(this);
```
писал, это надо реализовывать другим способом

alexstrive replied:
Удалено

andrsuh commented on src/main/resources/application.properties:
```diff
@@ -3,3 +3,8 @@ spring.data.neo4j.password=secret
 spring.kafka.consumer.group-id=supervisor
+
+
+management.endpoints.web.exposure.include=prometheus
```
проверил, что работает?

alexstrive replied:
Да

alexstrive committed:
remove metrics from exception constructors

alexstrive committed:
improve documention

alexstrive committed:
change naming

alexstrive committed:
rename all increase* methods to count*

andrsuh reviewed (dismissed):
Прекрасно!

Anglopol commented:
Пока не надо мерджить, я напишу когда

Anglopol reviewed (dismissed):
Можно мерджить, но перед этим подмерджи к себе дев, мы перешли на мавен

alexstrive committed:
Merge pull request #1 from microservices-course-itmo/dev

Dev

Anglopol committed:
run services in local docker (#12)

alexstrive committed:
Merge branch 'dev' into feature/prometheus

alexstrive committed:
solve conflict in application.properties

alexstrive commented:
Обновлено

alexstrive committed:
fixes

alexstrive committed:
Merge pull request #4 from microservices-course-itmo/dev

run services in local docker (#12)

alexstrive committed:
Merge branch 'fix/incorrect-image-and-kafka-address' into feature/prometheus

alexstrive committed:
remove redundant version
